extensions/applicationinsights-clickanalytics-js/src/common/Utils.ts

```diff
@@ -0,0 +1,362 @@
+/**
+ * @copyright Microsoft 2020
+ * File containing utility functions.
+ */
+
+import { CoreUtils, _InternalMessageId, hasDocument } from "@microsoft/applicationinsights-core-js";
+import {
+    IClickAnalyticsConfiguration
+} from '../Interfaces/Datamodel';
+
+
+const Prototype = 'prototype';
+const DEFAULT_DONOT_TRACK_TAG = 'ai-dnt';
+const DEFAULT_AI_BLOB_ATTRIBUTE_TAG = 'ai-blob';
+const DEFAULT_DATA_PREFIX = 'data-';
+
+export const _ExtendedInternalMessageId = {
+    ..._InternalMessageId,
+    CannotParseAiBlobValue: 101,
+    InvalidContentBlob: 102,
+    TrackPageActionEventFailed: 103
+};
+
+/**
+ * Finds attributes in overrideConfig which are invalid or should be objects
+ * and deletes them. useful in override config
+ * @param overrideConfig - override config object
+ * @param attributeNamesExpectedObjects - attributes that should be objects in override config object
+ */
+export function removeNonObjectsAndInvalidElements(overrideConfig: IClickAnalyticsConfiguration, attributeNamesExpectedObjects: Array<string>): void {
+    removeInvalidElements(overrideConfig);
+    for (var i in attributeNamesExpectedObjects) {
+        if (attributeNamesExpectedObjects.hasOwnProperty(i)) {
+            var objectName = attributeNamesExpectedObjects[i];
+            if (typeof overrideConfig[objectName] === 'object') {
+                removeInvalidElements(overrideConfig[objectName]);
+            } else {
+                delete overrideConfig[objectName];
+            }
+        }
+    }
+}
+
+/**
+ * Finds attributes in object which are invalid 
+ * and deletes them. useful in override config
+ * @param object Input object
+ */
+export function removeInvalidElements(object: Object): void {
+    /// Because the config object 'callback' contains only functions, 
+    /// when it is stringified it returns the empty object. This explains
+    /// the workaround regarding 'callback'
+    for (var property in object) {
+        if (!isValueAssigned(object[property]) ||
+            (JSON.stringify(object[property]) === '{}' && (property !== 'callback'))) {
+            delete object[property];
+        }
+    }
+}
+
+/**
+ * Checks if value is assigned to the given param.
+ * @param value - The token from which the tenant id is to be extracted.
+ * @returns True/false denoting if value is assigned to the param.
+ */
+export function isValueAssigned(value: any) {
+    /// <summary> takes a value and checks for undefined, null and empty string </summary>
+    /// <param type="any"> value to be tested </param>
+    /// <returns> true if value is null undefined or emptyString </returns>
+    return !(CoreUtils.isNullOrUndefined(value) || value === '');
+}
+
+/**
+ * Determines whether an event is a right click or not 
+ * @param evt - Mouse event 
+ * @returns true if the event is a right click 
+ */
+export function isRightClick(evt: any): boolean {
+    if ('which' in evt) { // Chrome, FF, ...
```

Comment:
nit: This if 'which' check looks like it could be extracted to a helper as it's repeated several times.